rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,849 @@
+algorithms requires implementing a custom plugin or requesting modifications
+from the upstream. The algorithms employed by `age` are not currently
+compatible with FIPS, which means configuring encrypted sessions while in FIPS
+mode will result in failed startup.
+
+In order to support key unwrapping directly within HSM and KMS systems, this
+RFD proposes a custom `age` plugin that uses RSA 4096 keypairs in combination
+with the decryption APIs exposed by each keystore backend. The choice of RSA is
+largely due to it's wide compatibility across all keystores supported by
+Teleport.
+
+Below is a high level diagram showing how `age` encryption and decryption work:
+![age diagram](assets/0127-age-high-level.png)
+
+### Config Changes
+
+Encrypted session recording is a feature of the auth service and can be enabled
+using a new `encryption` section in the `session_recording_config` resource.
+
+```yaml
+# session_recording_config.yml
+kind: session_recording_config
+version: v2
+spec:
+  encryption:
+    # whether or not encryption should be enabled
+    enabled: true
+    # whether or not Teleport should manage the keys or simply consume them
+    manual_key_management:
+      enabled: true
+      # key labels Teleport will use to find active encryption keys when
+      # 'enabled' is true
+      active_keys:
+        - type: pkcs11
+          label: 'pkcs11-key-label'
+        - type: aws_kms
+          label: 'aws key ID'
+      # key labels Teleport will use to find rotated decryption keys when
+      # 'enabled' is true
+      rotated_keys: []
+```
+
+HSM integration is facilitated through the existing configuration
+options for setting up an HSM backed CA keystore through pkcs#11. Example
+configuration found [here](https://goteleport.com/docs/admin-guides/deploy-a-cluster/hsm/#step-25-configure-teleport).
+
+The `manual_key_management` configuration controls whether or not the wrapping
+keys used by `age` should be provisioned and managed by Teleport or by the
+configured keystore. When disabled, Teleport will provision, share, and rotate
+keys as necessary. When enabled, Teleport expects to have access to keys
+queryable using the key labels assigned to `active_keys` and `rotated_keys`.
+Any keys found will be used during encryption and decryption and Teleport will
+not attempt to provision new keys or facilitate rotation.
+
+These configuration options are also accessible in the `teleport.yml` file
+configuration under the `auth_service` section:
+
+```yaml
+# teleport.yml
+auth_service:
+  session_recording: node
+  session_recording_config:
+    encryption:
+      enabled: true
+      manual_key_management:
+        enabled: true
+        active_keys:
+          - type: aws_kms
+            label: 'aws key ID'
+        rotated_keys: []
+```
+
+### Protobuf Changes
+
+```proto
+// api/proto/teleport/legacy/types/types.proto
+
+// EncryptionKeyPair is a keypair used for encrypting and decrypting data.
+message EncryptionKeyPair {
```

Comment:
EncryptionKeyPair doesn't need to be in api/proto/teleport/legacy/types/types.proto, right?